### PR TITLE
Redirect user after post

### PR DESCRIPTION
**Pull request by susiecoleman**
Currently if a user's session times out there is a 403 error and the user is directed to play's default 403 page. This PR overrides this behaviour and redirects to the login page with an error message for the user.

### CODE REVIEW

**Comment by danmassie** [comment]
What is causing the 403 to be returned? The AuthAction should return a 401 if unauthorized, and I can't see any reason for the default handling kicking in via the AuthAction, which makes me think that it must be a filter that is being invoked prior to the request reaching the action. The CSRF filter returns a 403 if the CSRF token is invalid, so I think this is where the error is coming from given that it only occurs on a POST.

Putting this blanket error handling in place does solve the immediate problem, but I'm wondering if this is actually down to misconfiguration.

File: app/views/login.scala.html [hunk 1]
```diff
@@ -18,6 +19,12 @@
                     We could not log you in. Please contact @email if the problem persists.
                 </div>
             }
+            case (Some(AccessForbidden()), _) => {
+                <div class="alert alert-danger" role="alert">
+                    We could not fulfill that request.
```
Comment (danmassie) [inline app/views/login.scala.html]: Why is the text hardcoded in this view rather than in the messages bundle?

**Comment by susiecoleman** [comment]
The 403 is returned because the expired session cookie is removed by the browser before making the request so the AuthAction has nothing to authenticate against. This doesn't matter for get requests but does for post @mchv might be better placed to explain why that is...

**Comment by danmassie** [comment]
If there is no session, then the userIdentity extracted from the request at https://github.com/guardian/play-googleauth/blob/master/module/src/main/scala/com/gu/googleauth/actions.scala#L68 will be None. This will result in the onUnauthorized handler, in this case https://github.com/guardian/play-googleauth/blob/master/module/src/main/scala/com/gu/googleauth/actions.scala#L51 to be invoked. This would not result in either a 401 or a 403. If the default onUnauthorized action of AuthenticatedBuilder was invoked, a 401 would be returned.

So I can't see how AuthAction is producing a 403.

**Comment by mchv** [comment]
@danmassie 
> The CSRF filter returns a 403 if the CSRF token is invalid, so I think this is where the error is coming from given that it only occurs on a POST.
This is correct. But the CSRF filter need the cookie to check if the provided token is valid.

**Comment by susiecoleman** [comment]
Just talking to James and Mariot it may be how we're expiring cookies that's causing the problem....? Looking at it now

File: app/views/login.scala.html (hunk 1 above)
Comment (susiecoleman) [inline app/views/login.scala.html]: Fixed

**Comment by susiecoleman** [comment]
403 problem has been fixed in another way. However this pull request has still achieved 2 things:
1. Removed text from login page and put it into messages.conf
2. Created custom responses to 400 errors so that the play defaults are not used.

File: app/services/CustomErrorHandler.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+class CustomErrorHandler @Inject() (
+                                   env: Environment,
+                                   config: Configuration,
+                                   sourceMapper: OptionalSourceMapper,
+                                   router: Provider[Router]
+                                   ) extends DefaultHttpErrorHandler(env, config, sourceMapper, router){
+
+  //Client Errors 4xx series
+  override def onBadRequest(request: RequestHeader, message: String): Future[Result] = {
```
Comment (danmassie) [inline app/services/CustomErrorHandler.scala]: BadRequest is used for invalid form submissions when updating users, so we cannot redirect to login.

Comment (susiecoleman) [inline app/services/CustomErrorHandler.scala]: So we're already handling it?

Comment (danmassie) [inline app/services/CustomErrorHandler.scala]: When we explicitly send a 400 for a form error, we provide the location (the view template) and display an error message, so in that situation it is handled. That's the only scenario in which a 400 can be encountered.

Comment (susiecoleman) [inline app/services/CustomErrorHandler.scala]: Removed

**Comment by danmassie** [comment]
:+1:

**Pull request merged**